### PR TITLE
notebooks: improved sharing UI

### LINKED ISSUE #30392: notebooks: add additional sharing capabilities
Currently, notebooks have two visibility settings: public or private. Public exposes the notebook to everyone on the instance (and logged-out users on Cloud). Meanwhile, a private notebook is only available to its creator.

This severely limits the sharing capabilities of notebooks on Cloud, where you would like to share notebooks that contain private code only within your organization.

Following the example of search contexts and batch changes, we would like to introduce namespaces to notebooks. Each notebook would belong to either a user namespace or an organization namespace.

### Summary of read and write permissions

User namespace
* Public notebooks: everyone can access the notebook, only the creator can edit it
* Private notebooks: only the creator can access the notebook, only the creator can edit it

Organization namespace
* Public notebooks: everyone can access the notebook, only organization members can edit it
* Private notebooks: only organization members can access it, only organization members can edit it

~~We have to decide if we want to handle changing the notebook namespace separately. If we follow the above rules, something like this could happen: a user (user A) creates a notebook under his namespace, then changes the namespace to an organization (org A) because he wants to share it with its members. Since everyone within the organization can edit the notebook, another organization member (user B) could transfer the notebook to his namespace and effectively "steal" the notebook from org A and user A. We could mitigate this by only allowing notebook creators to change notebook namespaces. In the previous example, only user A could change the namespace.~~

^ The above paragraph might be over-complicating things and I don't think its worth it to overcomplicate our permission for an unlikely scenario.

Work items:

- [x] Add namespace columns to the notebooks database table. Set the user namespace column to the creator column.
- [x] Update notebook database access permissions to account for new namespaces.
- [x] Update notebooks GQL API to return the namespace parameter for notebooks. Update write permissions check.
- [x] UI: add list tab for each organization.
- [x] UI: add [share modal clickable prototype](https://www.figma.com/proto/GU2dogrMzl0LkhwAzG3dUo/Notebooks-sharing?node-id=1714%3A32661&scaling=min-zoom&page-id=246%3A12&starting-point-node-id=1714%3A32661&show-proto-sidebar=1) which contains the current private/public visibility settings and the new namespaces selector. Source Figma is [here](https://www.figma.com/file/GU2dogrMzl0LkhwAzG3dUo/Notebooks-sharing?node-id=1714%3A26925)

### PR DESCRIPTION
**Pull request by novoselrok**
Fixes #30392

## Demo

https://user-images.githubusercontent.com/6417322/153391497-156842df-eac5-409c-abe4-d2f97e4fe976.mov

## Test plan

* Create a new notebook
* Open share dialog and change the share dropdown from 'Private' to an org, do not check the "Everyone can view..." checkbox, click done to apply the changes
* Go to the notebooks list page, click on the org tab and you should see the previously created notebook
* Login with a separate user that is also a member of the org
  * This user should be able to view & edit the created notebook, as well as view it in the org's notebooks list
* Login with a separate user that is *not* a member of the org
  * You should not be able to access the notebook, or view it in the explore list
* (In dotcom mode) Using an unauthenticated user, navigate to the created notebook
  * You should not be able to access the notebook, or view it in the explore list
* Go to the created notebook share dialog and check the "Everyone can view..." checkbox
* (In dotcom mode) Again, using an unauthenticated user, navigate to the notebook
  * You **should** be able to access the notebook

All read & write permissions are automatically tested at the GQL API layer and at the DB layer.

### CODE REVIEW

**Comment by novoselrok** [comment]
@camdencheek There was a small bug in the update logic, so there are some backend changes that have to be reviewed.

File: client/web/src/search/notebook/listPage/SearchNotebooksListPage.tsx [hunk 1]
```diff
@@ -66,20 +85,20 @@ export const SearchNotebooksListPage: React.FunctionComponent<SearchNotebooksLis
...
-    const [selectedTab, setSelectedTab] = useState<SelectedTab>(
+    const [selectedTab, setSelectedTab] = useState<NotebooksTab>(
         getSelectedTabFromLocation(location.search, authenticatedUser)
     )
 
     const setTab = useCallback(
-        (tab: SelectedTab) => {
+        (tab: NotebooksTab) => {
             setSelectedTab(tab)
             setSelectedLocationTab(location, history, tab)
```
Comment (fkling) [inline client/web/src/search/notebook/listPage/SearchNotebooksListPage.tsx:95]: I guess this logic could be moved to `onSelectTab`? That seems to be the only places that calls `setTab`.

**Review by fkling — APPROVED** [review]

**Review by camdencheek — APPROVED** [review]
Backend changes look good pending that one comment 👍

File: enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go [hunk 2]
```diff
@@ -387,6 +391,12 @@ func (r *notebookResolver) Namespace(ctx context.Context) (*graphqlbackend.Names
 	if r.notebook.NamespaceOrgID != 0 {
 		n, err := graphqlbackend.NamespaceByID(ctx, r.db, graphqlbackend.MarshalOrgID(r.notebook.NamespaceOrgID))
 		if err != nil {
+			// On Cloud, the user can have access to an org notebook if it is public. But if the user is not a member of
```
[the remark this answers is not in the document]
Comment (novoselrok) [reply enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go:393]: Made a slight modification after Camden's review: https://github.com/sourcegraph/sourcegraph/pull/30949#discussion_r803825968

[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go:393]: qq: on the opposite scenario, would the existence of an org be reveled when I look at the list of public notebooks or individual public notebook from orgs that I'm not a member of?

[the remark this answers is not in the document]
Comment (novoselrok) [reply enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go:393]: On the notebooks list page, you'll only get a tab for the orgs you are a member. We don't reveal the namespace (org) of the notebook anywhere in the list.

But if the notebook is public and the org is revealed in the contents of the notebook, then there is not much we can do.

[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go:393]: OK, SGTM

**Pull request merged**